src/gfn/gflownet/base.py

```diff
@@ -1,18 +1,23 @@
-from abc import abstractmethod
-from typing import Tuple
+from abc import ABC, abstractmethod
+from typing import Generic, Tuple, TypeVar, Union
 
 import torch
 import torch.nn as nn
 from torchtyping import TensorType as TT
 
 from gfn.containers import Trajectories
+from gfn.containers.base import Container
 from gfn.env import Env
 from gfn.modules import GFNModule
 from gfn.samplers import Sampler
 from gfn.states import States
 
+TrainingSampleType = TypeVar(
```

Comment:
This seems pretty general, I don't see any issues here (@saleml do you?)

Reply:
I like this! I believe the bound can be extended in case there are new losses that come up